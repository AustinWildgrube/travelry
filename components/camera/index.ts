export * from './Camera';
export * from './Permissions';