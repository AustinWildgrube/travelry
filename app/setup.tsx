import { Fragment, useState } from 'react';

import { Link } from 'expo-router';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Confetti, SetupProfileDetails, SetupProfileImage } from '&/components/auth';
import { Button, Heading, Text } from '&/components/core';
import { CONFETTI_FALL_SPEED, CONFETTI_ITEM_HEIGHT, CONFETTI_ITEM_WIDTH } from '&/utils/constants';
import { makeStyles } from '&/utils/makeStyles';

export default function Setup(): JSX.Element {
  const styles = useStyles();
  const [step, setStep] = useState(0);

  return (
    <SafeAreaView style={styles.setup}>
      <Heading>Setup your profile</Heading>

      {step === 0 ? (
        <Fragment>
          <Text style={styles.setupSlogan}>Upload an image - it's worth a thousand words</Text>
          <SetupProfileImage setStep={setStep} />

          <Button onPress={() => setStep(1)} variant="link" style={styles.skipLink}>
            <Button.Text>Skip</Button.Text>
          </Button>

          <Confetti
            fallSpeed={CONFETTI_FALL_SPEED}
            flipSpeed={3}
            horizSpeed={50}
            itemColors={['#84a98c', '#ffb703', '#023047']}
            itemDimensions={{ height: CONFETTI_ITEM_HEIGHT, width: CONFETTI_ITEM_WIDTH }}
            itemTintStrength={0.8}
            numItems={100}
          />
        </Fragment>
      ) : (
        <Fragment>
          <Text style={styles.setupSlogan}>Tell the world a little about yourself</Text>
          <SetupProfileDetails />

          <Link href="/feed" asChild>
            <Button variant="link" style={styles.skipLink}>
              <Button.Text>Skip</Button.Text>
            </Button>
          </Link>
        </Fragment>
      )}
    </SafeAreaView>
  );
}

const useStyles = makeStyles(theme => ({
  setup: {
    flex: 1,
    padding: theme.space['4'],
  },
  setupSlogan: {
    marginTop: theme.space['1'],
    marginBottom: theme.space['6'],
  },
  skipLink: {
    marginTop: theme.space['3'],
    zIndex: -1,
  },
}));
