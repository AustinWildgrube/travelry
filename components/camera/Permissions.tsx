import { Linking, Platform } from 'react-native';

import { ActivityAction, startActivityAsync } from 'expo-intent-launcher';

import { Button, Center, Heading, Text } from '&/components/core';

// TODO: once the camera is fixed check that this looks okay; it probably does not
export function Permissions(): JSX.Element {
  const openAppPref = (): void => {
    if (Platform.OS === 'ios') {
      Linking.openURL('app-settings:');
    } else {
      startActivityAsync(ActivityAction.SECURITY_SETTINGS);
    }
  };

  return (
    <Center>
      <Heading>Travelry needs access to your camera</Heading>
      <Text>Click the button below to enable it</Text>

      <Button>
        <Button.Text onPress={openAppPref}>Open Settings</Button.Text>
      </Button>
    </Center>
  );
}
