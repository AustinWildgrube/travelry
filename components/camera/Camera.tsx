import { useEffect, useRef, useState, type Dispatch, type SetStateAction } from 'react';
import { StyleSheet, View } from 'react-native';

import { Feather } from '@expo/vector-icons';
import { CameraType, Camera as ExpoCamera } from 'expo-camera';
import { type CameraCapturedPicture } from 'expo-camera/src/Camera.types';
import { FlipType, manipulateAsync } from 'expo-image-manipulator';
import { MediaTypeOptions, launchImageLibraryAsync, type ImagePickerAsset } from 'expo-image-picker';
import Animated, { SlideInDown, SlideOutUp } from 'react-native-reanimated';
import { SafeAreaView } from 'react-native-safe-area-context';

import { Permissions } from '&/components/camera/Permissions';
import { HStack, Pressable } from '&/components/core';
import { makeStyles } from '&/utils/makeStyles';

type CameraProps = {
  onSuccess: (imageUri: CameraCapturedPicture | ImagePickerAsset) => void;
  setIsShown: Dispatch<SetStateAction<boolean>>;
};

// TODO: figure out how to make full screen
export function Camera({ onSuccess, setIsShown }: CameraProps): JSX.Element {
  const styles = useStyles();
  const cameraRef = useRef<ExpoCamera | null>(null);

  const [hasCameraPermissions, requestCameraPermissions] = ExpoCamera.useCameraPermissions();
  const [type, setType] = useState(CameraType.front);

  const hideCamera = (): void => {
    setIsShown(false);
  };

  const flipCamera = (): void => {
    setType(type === CameraType.back ? CameraType.front : CameraType.back);
  };

  const pickImage = async (): Promise<void> => {
    let result = await launchImageLibraryAsync({
      mediaTypes: MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 3],
      quality: 1,
      exif: true,
    });

    if (!result.canceled) {
      onSuccess(result.assets[0]);
    }
  };

  const takePicture = async (): Promise<void> => {
    if (cameraRef && cameraRef.current instanceof ExpoCamera) {
      await cameraRef.current.takePictureAsync({ onPictureSaved: onPictureSaved, isImageMirror: false });
    }
  };

  const onPictureSaved = async (photo: CameraCapturedPicture): Promise<void> => {
    if (type === CameraType.front) {
      photo = await manipulateAsync(photo.uri, [{ rotate: 180 }, { flip: FlipType.Vertical }], {
        compress: 1,
      });
    }

    setIsShown(false);
    onSuccess(photo);
  };

  const requestCameraPermission = async (): Promise<void> => {
    if (!hasCameraPermissions) {
      await requestCameraPermissions();
    }
  };

  useEffect(() => {
    requestCameraPermission();
  }, []);

  return (
    <Animated.View entering={SlideInDown.duration(500)} exiting={SlideOutUp.duration(500)} style={styles.camera}>
      <ExpoCamera type={type} ref={cameraRef} style={styles.cameraContainer} testID="camera">
        <SafeAreaView style={styles.cameraSafeArea}>
          <HStack style={styles.cameraActionsTop}>
            <Pressable onPress={hideCamera} accessibilityLabel="Close camera">
              <Feather name="x" size={24} color="white" />
            </Pressable>

            <Pressable onPress={flipCamera} accessibilityLabel="Flip camera">
              <Feather name="refresh-cw" size={21} color="white" />
            </Pressable>
          </HStack>

          <HStack style={styles.cameraActionsBottom}>
            <Pressable onPress={pickImage} accessibilityLabel="Go to library" style={styles.cameraRollButton}>
              <Feather name="image" size={32} color="white" />
            </Pressable>

            <Pressable onPress={takePicture} accessibilityLabel="Capture image" style={styles.cameraCaptureRing}>
              <View style={styles.cameraCaptureButton} />
            </Pressable>
          </HStack>

          {!hasCameraPermissions && <Permissions />}
        </SafeAreaView>
      </ExpoCamera>
    </Animated.View>
  );
}

const useStyles = makeStyles(theme => ({
  camera: {
    ...StyleSheet.absoluteFillObject,
  },
  cameraContainer: {
    flex: 1,
  },
  cameraSafeArea: {
    flex: 1,
    marginBottom: theme.space['16'],
    marginHorizontal: theme.space['5'],
    marginTop: theme.space['5'],
  },
  cameraActionsTop: {
    flex: 1,
    justifyContent: 'space-between',
  },
  cameraActionsBottom: {
    alignItems: 'center',
    justifyContent: 'center',
  },
  cameraRollButton: {
    left: theme.space['0'],
    position: 'absolute',
  },
  cameraCaptureRing: {
    borderColor: theme.colors['white'],
    borderRadius: theme.radii['full'],
    borderWidth: theme.borderWidths['4'],
  },
  cameraCaptureButton: {
    backgroundColor: theme.colors['white'],
    borderRadius: theme.radii['full'],
    height: theme.space['20'],
    margin: theme.space['0.5'],
    width: theme.space['20'],
  },
}));
