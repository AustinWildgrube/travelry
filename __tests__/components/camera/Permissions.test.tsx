import { Linking } from 'react-native';

import { fireEvent, screen } from '@testing-library/react-native';

import { Permissions } from '&/components/camera/Permissions';

import { render } from '&/jest.setup';

describe('Camera Permissions', () => {
  beforeEach(() => {
    render(<Permissions />);
  });

  it('should match the snapshot', () => {
    screen.toJSON();
    expect(screen).toMatchSnapshot();
  });

  it('should render the component correctly', () => {
    expect(screen.getByText(/Travelry needs access to your camera/i)).toBeDefined();
    expect(screen.getByText(/Click the button below enable it/i)).toBeDefined();
    expect(screen.getByRole('button', { name: /Open Settings/i })).toBeDefined();
  });

  it('should open the app settings when "Open Settings" button is pressed', () => {
    const openSettingsButton = screen.getByRole('button', { name: /Open Settings/i });
    fireEvent.press(openSettingsButton);

    expect(Linking.openURL).toHaveBeenCalled();
  });
});
