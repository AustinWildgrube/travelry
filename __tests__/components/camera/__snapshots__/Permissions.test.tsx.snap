// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Camera Permissions should match the snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <View
    style={
      {
        "backgroundColor": "#ffffff",
        "flex": 1,
        "zIndex": 1,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.75)",
          "bottom": 0,
          "flex": 1,
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#fff",
            "fontFamily": "Inter_700Bold",
            "fontSize": 32,
            "fontWeight": "700",
            "lineHeight": 40,
            "textAlign": "center",
          }
        }
      >
        Travelry needs access to your camera
      </Text>
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#fff",
            "fontFamily": "Inter_400Regular",
            "fontSize": 16,
            "fontWeight": "400",
            "lineHeight": 20,
            "marginBottom": 24,
            "marginTop": 8,
            "textAlign": "center",
          }
        }
      >
        Click the button below enable it
      </Text>
      <View
        style={
          [
            {
              "overflow": "hidden",
            },
            {
              "borderRadius": 8,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            {
              "busy": false,
              "checked": undefined,
              "disabled": false,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#84a98c",
                "borderBottomWidth": 4,
                "borderColor": "#6A8770",
                "borderRadius": 8,
                "borderStyle": "solid",
                "borderWidth": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 12,
                "paddingHorizontal": 14,
              }
            }
          >
            <Text
              style={
                {
                  "color": "#fff",
                  "fontSize": 18,
                  "fontWeight": "600",
                  "paddingVertical": 1,
                  "textAlign": "center",
                }
              }
            >
              Open Settings
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
